src/language-server/config/config.ts

```diff
@@ -1,146 +1,225 @@
 import { dirname } from "path";
-import merge from "lodash.merge";
-import { ClientID, ServiceID, ServiceSpecifier } from "../engine";
 import { URI } from "vscode-uri";
-import { WithRequired } from "../../env";
 import { getGraphIdFromConfig, parseServiceSpecifier } from "./utils";
+import { Debug } from "../utilities";
+import z, { ZodError } from "zod";
 import { ValidationRule } from "graphql/validation/ValidationContext";
+import { Slot } from "@wry/context";
+import { fromZodError } from "zod-validation-error";
 
-export interface EngineStatsWindow {
-  to: number;
-  from: number;
+const ROVER_AVAILABLE = (process.env.APOLLO_FEATURE_FLAGS || "")
+  .split(",")
+  .includes("rover");
+
+function ignoredFieldWarning(
+  getMessage = (path: string) =>
+    `The option ${path} is no longer supported, please remove it from your configuration file.`,
+) {
+  return z
+    .custom(() => true)
+    .superRefine((val, ctx) => {
+      if (val) {
+        Debug.warning(getMessage(ctx.path.join(".")));
+      }
+    })
+    .transform((val) => {
+      return undefined;
+    });
+}
+export interface Context {
+  apiKey?: string;
+  serviceName?: string;
 }
+const context = new Slot<Context>();
 
-export const DefaultEngineStatsWindow = {
-  to: -0,
-  from: -86400, // one day
-};
+const studioServiceConfig = z.string();
 
-export interface HistoricalEngineStatsWindow extends EngineStatsWindow {}
+const remoteServiceConfig = z.object({
+  name: z.string().optional(),
+  url: z.string(),
+  headers: z.record(z.string()).default({}),
+  skipSSLValidation: z.boolean().default(false),
+});
+export type RemoteServiceConfig = z.infer<typeof remoteServiceConfig>;
 
-export type EndpointURI = string;
-export interface RemoteServiceConfig {
-  name: ServiceID;
-  url: EndpointURI;
-  headers?: { [key: string]: string };
-  skipSSLValidation?: boolean;
-}
+const localServiceConfig = z.object({
+  name: z.string().optional(),
+  localSchemaFile: z.union([z.string(), z.array(z.string())]),
+});
+export type LocalServiceConfig = z.infer<typeof localServiceConfig>;
 
-export interface LocalServiceConfig {
-  name: ServiceID;
-  localSchemaFile: string | string[];
-}
+const clientServiceConfig = z.preprocess(
+  (value) => value || context.getValue()?.serviceName,
+  z.union([studioServiceConfig, remoteServiceConfig, localServiceConfig]),
+);
+export type ClientServiceConfig = z.infer<typeof clientServiceConfig>;
 
-export interface EngineConfig {
-  endpoint?: EndpointURI;
-  readonly apiKey?: string;
-}
+const clientConfig = z.object({
+  service: clientServiceConfig,
+  validationRules: z
+    .union([
+      z.array(z.custom<ValidationRule>()),
+      z.function().args(z.custom<ValidationRule>()).returns(z.boolean()),
+    ])
+    .optional(),
```

Comment:
I don't think this feature has ever been documented anywhere, but it's fully implemented, so I'm not dropping it now.